Feature/net8 regex

gerhardol opened the pull request:
## Proposed changes

- Enable C# 12

- https://devblogs.microsoft.com/dotnet/regular-expression-improvements-in-dotnet-7/#source-generation

Compile all static regex at compile time instead of at runtime.
See also #11370 that refers to https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-regex#regular-expressions-compiled-to-an-assembly
Before .NET7 either interpreted that executes slightly slower or compiled that slows down starting was used.  
For .NET7 the recommended use is to compile regex at compile time, with both advantages.

This PR implements the analyzer suggestion with a few manual tweaks:
- For all regex with previous `private static readonly Regex` the `[GeneratedRegex` attribute was placed in the top, to simplify comparisons
- Option RegexOptions.Compiled was manually removed
- The default name MyRegex was changed to something meaningful
- Some variables without any use was removed
- A few minor adjustments and suggestions done.

## Test methodology <!-- How did you ensure quality? -->

- Tests still run

## Merge strategy

- To be decided later.

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented on Plugins/BuildServerIntegration/AzureDevOpsIntegration/Settings/ProjectUrlHelper.cs:
```diff
@@ -174,5 +172,26 @@ public static (bool success, string? projectUrl, int buildId) TryParseBuildUrl(s
+        [GeneratedRegex(@"^(?<prot>(?:http|https))://(?<user>[^.@]+)(?:@[^.]*)?\.visualstudio\.com(?<port>:\d*)?(?:/DefaultCollection)?(?<project>(/[^/]+)?/[^/]+)/_(git|ssh)/(.+)$")]
+        private static partial Regex VsTeamHttpsRegex();
+        [GeneratedRegex(@"^(?<user>[^.@]+)@vs-ssh\.visualstudio.com:v3(?:/[^/]*)?(?<project>/[^/]+)")]
+        private static partial Regex VsTeamHttpsSsh();
```
```suggestion
        private static partial Regex VsTeamHttpsSshRegex();
```

gerhardol replied:
Both Regex changed

gerhardol commented on Externals/NetSpell.SpellChecker/Spelling.cs:
```diff
@@ -17,16 +17,22 @@ namespace NetSpell.SpellChecker
-        private readonly Regex _htmlRegex = new(@"</[c-g\d]+>|</[i-o\d]+>|</[a\d]+>|</[q-z\d]+>|<[cg]+[^>]*>|<[i-o]+[^>]*>|<[q-z]+[^>]*>|<[a]+[^>]*>|<(\[^\]*\|'[^']*'|[^'\>])*>", RegexOptions.IgnoreCase & RegexOptions.Compiled);
+        [GeneratedRegex(@"^\d")]
+        private static partial Regex DigitRegex();
+        [GeneratedRegex(@"</[c-g\d]+>|</[i-o\d]+>|</[a\d]+>|</[q-z\d]+>|<[cg]+[^>]*>|<[i-o]+[^>]*>|<[q-z]+[^>]*>|<[a]+[^>]*>|<(\[^\]*\|'[^']*'|[^'\>])*>")]
```
This changes the behavior though: `&`
But probably what was intended...

gerhardol commented on GitUI/UserControls/ConsoleEmulatorOutputControl.cs:
```diff
@@ -184,7 +184,7 @@ private void SendAsLines(string output)
             }
 
-            string[] outputLines = Regex.Split(output, @"(?<=[\n\r])");
+            string[] outputLines = MyRegex().Split(output);
```
That is the default name - most work with this PR was to rename all MyRegex1 etc
(I reverted Conemu changes but this need to be changed...)

mstv commented on Plugins/BuildServerIntegration/AzureDevOpsIntegration/Settings/ProjectUrlHelper.cs:
```diff
@@ -6,54 +6,52 @@ namespace AzureDevOpsIntegration.Settings
             { // VS Team Services via HTTPS
-                new Regex(@"^(?<prot>(?:http|https))://(?<user>[^.@]+)(?:@[^.]*)?\.visualstudio\.com(?<port>:\d*)?(?:/DefaultCollection)?(?<project>(/[^/]+)?/[^/]+)/_(git|ssh)/(.+)$"),
+                VsTeamSshRegex(),
                 (match) => $"{match.Groups["prot"].Value}://{match.Groups["user"].Value}.visualstudio.com{match.Groups["port"].Value}{match.Groups["project"].Value}"
             },
             { // VS Team Services via SSH
-                new Regex(@"^(?<user>[^.@]+)@vs-ssh\.visualstudio.com:v3(?:/[^/]*)?(?<project>/[^/]+)"),
+                VsTeamHttpsRegex(),
```
Swap back, please.

gerhardol replied:
Thanks!

mstv commented on Plugins/BuildServerIntegration/AzureDevOpsIntegration/Settings/ProjectUrlHelper.cs:
```diff
@@ -174,5 +172,26 @@ public static (bool success, string? projectUrl, int buildId) TryParseBuildUrl(s
             return (false, "", -1);
         }
+
+        [GeneratedRegex(@"^(?<prot>(?:http|https))://(?<user>[^.@]+)(?:@[^.]*)?\.visualstudio\.com(?<port>:\d*)?(?:/DefaultCollection)?(?<project>(/[^/]+)?/[^/]+)/_(git|ssh)/(.+)$")]
+        private static partial Regex VsTeamSshRegex();
+        [GeneratedRegex(@"^(?<user>[^.@]+)@vs-ssh\.visualstudio.com:v3(?:/[^/]*)?(?<project>/[^/]+)")]
+        private static partial Regex VsTeamHttpsRegex();
```
also here

RussKie commented on BugReporter/BugReportForm.cs:
```diff
@@ -243,5 +243,8 @@ public TestAccessor(BugReportForm form)
             public static bool CheckContainsInfo(string input) => BugReportForm.CheckContainsInfo(input);
         }
+
+        [GeneratedRegex(@"\s|\r|\n")]
+        private static partial Regex WhitespaceRegex();
```
Let's keep these automagically generated methods at the top of the class after fields.
Ditto other instances.

gerhardol replied:
All generated occurrences to be changed, including future conversions then.

gerhardol replied:
Changed

gerhardol commented:
Rebased on master with the merged .net8. Squashed comments (and removed some not yet used GeneratedRegex for grep).
One open question is if `[GeneratedRegex` should be aligned and put first (as previous static readonly where that existed) or last (as when converted).

RussKie commented on GitUI/CommandsDialogs/AboutBoxDialog/FormContributors.cs:
```diff
@@ -34,12 +34,11 @@ void InitialiseComponent()
-                    _contributors.Text, Regex.Replace(Resources.Coders, NEWLINES, " "));
-                textBoxes[1].Text = Regex.Replace(Resources.Translators, NEWLINES, " ");
-                textBoxes[2].Text = Regex.Replace(Resources.Designers, NEWLINES, " ");
+                    _team.Text, NewlineRegex().Replace(Resources.Team, " "),
+                    _contributors.Text, NewlineRegex().Replace(Resources.Coders, " "));
+                textBoxes[1].Text = NewlineRegex().Replace(Resources.Translators, " ");
+                textBoxes[2].Text = NewlineRegex().Replace(Resources.Designers, " ");
```
Can we use a variable instead of multiple invocations?

gerhardol replied:
Optimizing one function call does not make any difference, just one more line of code.

```
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Text.RegularExpressions.Generator", "8.0.9.3103")]
        private static partial global::System.Text.RegularExpressions.Regex NewlineRegex() => global::System.Text.RegularExpressions.Generated.NewlineRegex_1.Instance;

...
    [GeneratedCodeAttribute("System.Text.RegularExpressions.Generator", "8.0.9.3103")]
    [SkipLocalsInit]
    file sealed class NewlineRegex_1 : Regex
    {
        /// <summary>Cached, thread-safe singleton instance.</summary>
        internal static readonly NewlineRegex_1 Instance = new();
```

RussKie reviewed (commented):
👍

gerhardol commented:
Rebased, moved GeneratedRegex to start of class with other private (a few regex are public too)

RussKie reviewed (approved):
Looks like some of the Verify snapshots need to be updated due to callstack changes.

gerhardol commented:
squashed and rebased
Fixed the tests. Just adding [GeneratedRegex] attribute changes the test results, not related to the actual test.

Will merge tomorrow unless there are protests
